[오승하] Add: filter 기능 구현

lilyoh opened the pull request:
## 수정 사항 간략한 한줄 요약
**2020-08-12**
Add: react monster 과제 완료

**2020-08-13**
Add: filter 기능 구현
		 
## 수정 사항들 자세한 내용
**2020-08-12**
- api 호출하여 데이터 로딩
- api 호출 결과값을 자식 컴포넌트에 전달
- map 사용하여 특정 값을 가진 컴포넌트 생성

**2020-08-13**
- search box 에 몬스터 이름 입력 시 몬스터 카드가 필터링되어 화면에 출력되는 기능 구현
- filter, toLowerCase, includes 사용

## 기타 질문 및 특이 사항

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [ ] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

jamessoun93 commented on src/2-monsters/index.js:
```diff
@@ -23,21 +23,37 @@ import "./index.css";
     // 필터링 로직
-
+    let newArr = this.state.monsters.filter((x) => {
+      return x.name.toLowerCase().includes(this.state.userInput);
```
더욱 더 정확한 비교를 위해 userInput 도 소문자 형태로 변환해주세요! 

그리고 filter 함수 내 callback 이 인자로 받는 객체가 monsters 배열 내에 있는 각각 monster 정보를 담고 있는 객체라서 조금 더 알아보기 쉬운 이름으로 변경해주셨으면 좋겠습니다 :)

Mingyu-Song replied:
넵! 수정하겠습니다.

lilyoh committed:
Add: react monster 과제 완료
- api 호출하여 데이터 로딩
- api 호출 결과값을 자식 컴포넌트에 전달
- map 사용하여 특정 값을 가진 컴포넌트 생성

lilyoh committed:
Add: filter 기능 구현
- search box 에 몬스터 이름 입력 시 몬스터 카드가 필터링되어 화면에
  출력되는 기능 구현
- filter, toLowerCase, includes 사용

jamessoun93 commented on src/Assignments/Monsters/students/11th/seungha-oh/Components/Card/Card.js:
```diff
@@ -22,11 +24,11 @@ class Card extends Component {
-          src="https://live.staticflickr.com/4561/38054606355_26429c884f_b.jpg"
+          src={`https://robohash.org/${this.props.id}?set=set2&size=180x180`}
           alt=""
         />
-        <h2>Name</h2>
-        <p>Email</p>
+        <h2>{this.props.name}</h2>
+        <p>{this.props.email}</p>
```
좋습니다 좋아요 😎

jamessoun93 commented on src/Assignments/Monsters/students/11th/seungha-oh/Components/CardList/CardList.js:
```diff
@@ -1,17 +1,29 @@
+  constructor() {
+    super();
+    this.state = {};
+  }
```
state 로 관리하고 싶은 데이터가 없다면 해당 constructor 전체가 불필요합니다! 😇

lilyoh replied:
삭제하였습니다.

jamessoun93 commented on src/Assignments/Monsters/students/11th/seungha-oh/Components/CardList/CardList.js:
```diff
@@ -1,17 +1,29 @@
-    return <div className="card-list"></div>;
+    return (
+      <div className="card-list">
+        {this.props.monsters.map((el, id) => {
+          return <Card key={id} id={el.id} name={el.name} email={el.email} />;
```
`map()` 너무 잘쓰셨네요 👍 👍 👍

jamessoun93 commented on src/Assignments/Monsters/students/11th/seungha-oh/Components/SearchBox/SearchBox.js:
```diff
@@ -1,4 +1,5 @@
 import React, { Component } from "react";
+import "../../Monsters";
```
음.. 이걸 여기에 넣으신 이유가 있을까요? 🤔

lilyoh replied:
이유없이 들어가 있는 것 같습니다. 삭제했습니다!

jamessoun93 commented on src/Assignments/Monsters/students/11th/seungha-oh/Monsters.js:
```diff
@@ -21,21 +20,38 @@ import "./Monsters.scss";
   // 데이터 로딩
-
+  componentDidMount() {
+    fetch("https://jsonplaceholder.typicode.com/users")
+      .then((res) => res.json())
+      .then((res) => this.setState({ monsters: res, filteredMonsters: res }));
```
fetch 함수 사용 완벽합니다 👍 👍

jamessoun93 commented on src/Assignments/Monsters/students/11th/seungha-oh/Monsters.js:
```diff
@@ -21,21 +20,38 @@ import "./Monsters.scss";
+      .then((res) => this.setState({ monsters: res, filteredMonsters: res }));
+  }
   // SearchBox 에 props로 넘겨줄 handleChange 메소드 정의
+  handleChange = (e) => {
+    this.setState({
+      userInput: e.target.value,
+      filteredMonsters: this.state.monsters.filter((monster) =>
+        monster.name.toLowerCase().includes(e.target.value.toLowerCase())
+      ),
+    });
+  };
```
아까 저랑 같이 한 부분이죠? ㅎㅎ 꼭 기억하고 계셔야 합니다! 🙇🏻‍♂️

jamessoun93 commented on src/Assignments/Monsters/students/11th/seungha-oh/Monsters.js:
```diff
@@ -21,21 +20,38 @@ import "./Monsters.scss";
         <h1>컴포넌트 재사용 연습!</h1>
-        {/* <SearchBox handleChange=정의한메소드 /> */}
-        {/* <CardList monsters=몬스터리스트 /> */}
+        <SearchBox handleChange={this.handleChange} />
+        <CardList monsters={this.state.filteredMonsters} />
```
👍 👍 👍

jamessoun93 reviewed (changes requested):
승하님 전체적으로 너무 깔끔해서 보는 내내 기분이 좋았어요 🥳 🥳 🥳 

몇가지 리뷰사항만 반영해서 새로 push 해주세요!!

lilyoh committed:
Refactor: 리뷰 반영하여 코드 수정
- CardList.js 파일 내의 불필요한 constructor 삭제
- SearchBox.js 파일 내의 불필요한 import 삭제

jamessoun93 commented:
LGTM!

수고하셨습니다~ 😎
